nni/algorithms/compression/v2/pytorch/pruning/tools/task_generator.py

```diff
@@ -89,12 +111,223 @@ def generate_config_list(self, target_sparsity: List[Dict], iteration: int, mode
+
+    def _adjust_target_sparsity(self):
+        """
+        If origin_masks is not empty, then re-scale the target sparsity.
+        """
+        if len(self.masked_rate) > 0:
+            for config in self.target_sparsity_list:
+                sparsity, op_names = config['total_sparsity'], config['op_names']
+                remaining_weight_numel = 0
+                pruned_weight_numel = 0
+                for name in op_names:
+                    remaining_weight_numel += self.weights_numel[name]
+                    if name in self.masked_rate:
+                        pruned_weight_numel += 1 / (1 / self.masked_rate[name] - 1) * self.weights_numel[name]
+                config['total_sparsity'] = max(0, sparsity - pruned_weight_numel / (pruned_weight_numel + remaining_weight_numel))
+
+    def _init_temp_config_list(self):
+        self._temp_config_list = []
+        self._temp_sparsity_list = []
+        for config in self.target_sparsity_list:
+            sparsity_config, sparsity = self._init_config_sparsity(config)
+            self._temp_config_list.extend(sparsity_config)
+            self._temp_sparsity_list.append(sparsity)
+
+    def _init_config_sparsity(self, config: Dict) -> Tuple[List[Dict], List]:
+        assert 'total_sparsity' in config, 'Sparsity must be set in config: {}'.format(config)
+        target_sparsity = config['total_sparsity']
+        op_names = config['op_names']
+
+        if target_sparsity == 0:
+            return [], []
+
+        while True:
+            random_sparsity = sorted(np.random.uniform(0, 1, len(op_names)))
+            rescaled_sparsity = self._rescale_sparsity(random_sparsity, target_sparsity, op_names)
+            if rescaled_sparsity is not None and rescaled_sparsity[0] >= 0 and rescaled_sparsity[-1] < 1:
+                break
+
+        return self._sparsity_to_config_list(rescaled_sparsity, config), rescaled_sparsity
+
+    def _rescale_sparsity(self, random_sparsity: List, target_sparsity: float, op_names: List) -> List:
+        assert len(random_sparsity) == len(op_names)
+
+        num_weights = sorted([self.weights_numel[op_name] for op_name in op_names])
+        sparsity = sorted(random_sparsity)
+
+        total_weights = 0
+        total_weights_pruned = 0
+
+        # calculate the scale
+        for idx, num_weight in enumerate(num_weights):
+            total_weights += num_weight
+            total_weights_pruned += int(num_weight * sparsity[idx])
+        if total_weights_pruned == 0:
+            return None
+
+        scale = target_sparsity / (total_weights_pruned / total_weights)
+
+        # rescale the sparsity
+        sparsity = np.asarray(sparsity) * scale
+        return sparsity
+
+    def _sparsity_to_config_list(self, sparsity: List, config: Dict) -> List[Dict]:
+        sparsity = sorted(sparsity)
+        op_names = [k for k, _ in sorted(self.weights_numel.items(), key=lambda item: item[1]) if k in config['op_names']]
```

Comment:
what is the meaning of this line?

Reply:
sort the `op_name` in `config['op_names']` with the number of op weight elements.